addons/l10n_in/models/account_invoice.py

```diff
@@ -71,18 +71,36 @@ def _compute_l10n_in_state_id(self):
 
     @api.depends('l10n_in_state_id')
     def _compute_fiscal_position_id(self):
-        other_country = None
-        in_moves = self.filtered(lambda move: move.country_code == 'IN' and move.is_sale_document(include_receipts=True))
+        foreign_state = False
+        in_moves = self.filtered(lambda move: move.country_code == 'IN' and move.is_invoice(include_receipts=True))
         for move in in_moves:
-            state_id = move.l10n_in_state_id.id
-            country_id = move.l10n_in_state_id.country_id.id
-            if move.l10n_in_state_id.l10n_in_tin == '96':  # Foreign Country
-                other_country = other_country or self.env['res.country'].search([('code', '!=', 'IN')], limit=1)
-                country_id = other_country.id
-                state_id = False
+            state_id = move.l10n_in_state_id
+            is_purchase_document = move.is_purchase_document(include_receipts=True)
+            if (
+                is_purchase_document
+                and state_id.l10n_in_tin != '96'
+                and move.partner_id.country_id.code == 'IN'
+            ):
+                if state_id == move.partner_id.state_id:
+                    # Set Intra State for purchase if place of supply is same as partner state
+                    state_id = move.company_id.state_id
```

Comment:
But partner state could still be for in another state than the company one.  Why do not we have to apply interstate in that case?

Reply:
Because here we are receiving the bill. (It's same as we do for the invoice but we are seeing the things from vendor respective that's the reason it should apply Intra)

Suppose the company is located in `Maharashtra` and Vendor located in `Gujarat`, Suppose delivery of goods are taken in warehouse\branch (i.e. Shipping address) located in `Gujarat`, the Vendor bill will have a Place of Supply as `Gujarat`. From Vendor's Perspective the transaction will be Intra State and will apply `GST x%`

So if it's a purchase document and the POS is same as vendor's state. It should be Intra state